### PR TITLE
Periodic and mapbackeds

### PR DESCRIPTION
**Pull request by kaifox**
Dear all,
please have a look... as discussed this contains a first version of the mapbackeds concept (at the moment quite lenient ... might need a few more consistency checks in the future)

and a simple periodically emitting obersvable.

### CODE REVIEW

File: src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            return toString();
+        }
+        if ("hashCode".equals(method.getName()) && (args == null)) {
+            return hashCode();
+        }
+        if ("equals".equals(method.getName()) && (args.length == 1)) {
+            return equals(args[0]);
+        }
+
+        throw new IllegalArgumentException(
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java:57]: [cosmetics] Maybe rather UnsupportedOperationException?

File: src/main/java/org/ossgang/commons/mapbackeds/Mapbackeds.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    public static class Builder<M> {
+
+        private final Class<M> backedInterface;
+        private final Set<Method> fieldMethods;
+        private final Map<String, Object> mapBuilder = new HashMap<>();
+
+        /*
+         * TODO: Introduce consistency checks
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/mapbackeds/Mapbackeds.java:126]: Are these TODOs still to do? at least some checks seem to be here?

File: src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Creates an observable value that emits periodically at the given rate. The value emitted is the current time (as
+ * instant). Emitting is done on a single thread.
+ */
+public class PeriodicObservableValue extends DispatchingObservableValue<Instant> {
+
+    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+
+    PeriodicObservableValue(long period, TimeUnit unit) {
+        super(null);
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java:17]: Initialize with Instant.now(), just to guarantee that a value is present at all time?

File: src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import java.time.Instant;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Creates an observable value that emits periodically at the given rate. The value emitted is the current time (as
+ * instant). Emitting is done on a single thread.
+ */
+public class PeriodicObservableValue extends DispatchingObservableValue<Instant> {
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java:12]: Perhaps add at least some basic test?

**Comment by michi42** [comment]
Looks good to me, with a few minor comments

File: src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+ */
+class MapbackedObject implements InvocationHandler {
+
+    private final Class<?> intfc;
+    private final Map<String, Object> fieldValues;
+    private final Set<Method> fieldMethods;
+
+    MapbackedObject(Class<?> intfc, Map<String, Object> fieldValues) {
+        this.intfc = requireNonNull(intfc, "interface must not be null");
+        this.fieldValues = Map.copyOf(fieldValues);
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java:32]: This is java 11 only ;-)

File: src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java (hunk 1 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java:57]: Agreed. Changed

File: src/main/java/org/ossgang/commons/mapbackeds/Mapbackeds.java (hunk 2 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/mapbackeds/Mapbackeds.java:126]: It is anyhow not nice to have those todos in the coed. Therefore, I created a new issue with the ideas #47  and removed the todos from the code.

File: src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java (hunk 3 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java:17]: Good idea. Done

File: src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java (hunk 5 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/mapbackeds/MapbackedObject.java]: fixed (copying to new HashMap in constructor and getter.

File: src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java (hunk 4 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/PeriodicObservableValue.java:12]: Added one (though vbery basic ;-)

**Pull request merged**